feat(storage): Clickhouse Query placeholder

fpacifici opened the pull request:
This is the first step in splitting logical query and storage (clickhouse) query.

It does the bare minimum to address two problems that I think would become much bigger if we start working together to the query planning: the query class used after storage selection is still the logical Snuba query, and the instance of the query object is still the same.
So this PR does the following:
- Introduce the Clickhouse Query class (which is really just a type alias to the old class). This allows every interface in the query processing pipeline not to depend on the logical query object after the storage has been selected.
- It copies the query during the translation phase, so that after storage selection we are not using the same mutable object anymore. Any change to the logical query at that point won't produce any effect.

- It also introduces a trivial QueryTranslator just to show when translation happens and to make it simpler to build on top of it, but that class is really in early stages.
- it renames everything storage specific into clickhouse specific.

fpacifici committed:
Add physical query placeholder

fpacifici committed:
Add translator

fpacifici committed:
Move the translator instantiation to the selector

fpacifici committed:
Fix some comments

fpacifici commented on snuba/clickhouse/astquery.py:
```diff
@@ -19,7 +19,7 @@ class AstClickhouseQuery(ClickhouseQuery):
     """
 
-    def __init__(self, query: Query, settings: RequestSettings,) -> None:
+    def __init__(self, query: PhysicalQuery, settings: RequestSettings,) -> None:
```
I called it PhysicalQuery since we are talking about physical model and logical model. Really no strong opinion on this. If someone prefers StorageQuery I will change it.

lynnagara replied:
We have some references to StorageQuery in other places (e.g. StorageQuerySplitStrategy, StorageQueryPlanBuilder) so it might be a bit more consistent.

tkaemming replied:
No strong preference here. I agree that consistency is good regardless of which 

 A disadvantage of `StorageQuery` is the implication that the query is run _on_ the storage rather than on the `Cluster` in coordination with the storage(s). Also "storage" being singular might send the wrong message as well since a query can reference multiple storages.

I think in an ideal scenario we would split the logical and physical domains in code and have `snuba.logical.Query` and `snuba.physical.Query`. The only thing that bridges between them would be the translation of `logical.Query` to `physical.Query` and everything else would be unambiguous by its location in the code: `snuba.logical.*` would reference the `logical.Query` exclusively, and similar for the physical query respectively. I am not sure things are clear cut enough yet to justify that scope of a change yet but that's the direction I'd like to see.

fpacifici replied:
makes sense. So that means Physical over Storage (whether it is in the class name or in the module name).
@lynnagara I think we can easily rename StorageQueryPlanBuilder. That came before we started talking about logical and physical.

fpacifici replied:
This should be addressed now:
- physical = clickhouse
- we have a logical query and a clickhouse query (same Query name in different packages)
- we have a ClickhouseQueryPlan and ClickhouseQueryPlanBuilder

fpacifici commented on snuba/datasets/discover.py:
```diff
@@ -89,11 +90,16 @@ def __init__(
             query, self.__abstract_events_columns, self.__abstract_transactions_columns,
         )
-        return self.__events_table if table == EVENTS else self.__transactions_table
+        translator = CopyTranslator()
```
Obviously the translation will depend on the storage we select, and the selection itself may end up being the result of the translation (depending how we decide to couple translation and storage).

fpacifici commented on snuba/datasets/plans/single_storage.py:
```diff
@@ -46,14 +54,16 @@ def __init__(
-        # request object to be moved around, thus the Query has to be mutable and cannot
-        # be simply rebuilt and replaced.
-        request.query.set_data_source(
+        # TODO: Clearly the QueryTranslator instance  will be dependent on the storage.
+        # Setting the data_source on the query should become part of the translation
+        # as well.
```
The main question here, that I think is besides the point of this PR since both solutions are viable, is whether:
1) the storage will provide a translator interface (which hopefully an instance of a general class that is capable to translate any schema into any schema)
2) We will assign here a translator that we build here starting from the schema of the entity and the schema of the storage
3) the storage just provides a "translate" method (which is probably going to be the case for the multi-table storages. 

I think in any of these cases there is likely going to be a translator interface that takes a query and spits a physical query.

fpacifici replied:
This has been discussed offline. The translator will be a visitor.

fpacifici commented on snuba/datasets/plans/single_storage.py:
```diff
@@ -71,18 +81,25 @@ class SelectedStorageQueryPlanBuilder(StorageQueryPlanBuilder):
     def build_plan(self, request: Request) -> StorageQueryPlan:
-        storage = self.__selector.select_storage(request.query, request.settings)
-        request.query.set_data_source(
+        storage, translator = self.__selector.select_storage(
```
I hope to make this work for joins as well, allowing us to pick a storage per entity that we are joining and a translator for each so that each translator acts independently on the joined query translating the part related to the storage the translator is coupled with. Will do some experiments around this.

lynnagara commented on snuba/query/physical.py:
```diff
@@ -0,0 +1,5 @@
+from snuba.query.query import Query
+
+# This defines the type for the physical query that operates on storages.
+# Obviously this is supposed to become more complex than just a type alias.
```
What are the differences that we'd expect to see here between the physical and logical query class?

fpacifici replied:
- clickhouse specific stuff like pre-where would not be in the logical query
- tags as mapping columns `tags[asdasd]` are very likely not going to be in the physical query.

There may be more.

lynnagara commented on snuba/datasets/plans/translators.py:
```diff
@@ -0,0 +1,31 @@
+from abc import ABC, abstractmethod
+
+from snuba.query.physical import PhysicalQuery
+from snuba.query.query import Query
```
Should this be named LogicalQuery now?

tkaemming replied:
+1

fpacifici committed:
Merge branch 'master' into feat/split_query_class

fpacifici committed:
Move physical and logical query in separate modules

fpacifici commented:
Now there are a logical module and a physical module in the query module. Classes are both named Query. Same for QueryProcessors

fpacifici commented:
WIP because I still have to rename the StorageQueryPlanBuilder

fpacifici commented on snuba/datasets/plans/translators.py:
```diff
@@ -0,0 +1,31 @@
+from snuba.query.logical import Query as LogicalQuery
+
+
+class QueryTranslator(ABC):
```
This is really just a placeholder at this point so that we are mindful that there is an explicitly called translation phase in query plan building. I expect the implementations of this class to evolve a lot after we nail down the design of the query translation

tkaemming commented on snuba/datasets/plans/single_storage.py:
```diff
@@ -7,19 +7,27 @@
     StorageQueryPlanBuilder,
 )
 from snuba.datasets.storage import QueryStorageSelector, ReadableStorage
+from snuba.datasets.plans.translators import CopyTranslator
 
 # TODO: Importing snuba.web here is just wrong. What's need to be done to avoid this
 # dependency is a refactoring of the methods that return RawQueryResult to make them
 # depend on Result + some debug data structure instead. Also It requires removing
 # extra data from the result of the query.
 from snuba.web import RawQueryResult
-from snuba.query.query_processor import QueryProcessor
+from snuba.query.physical import Query
+from snuba.query.processors.physical import QueryProcessor
 from snuba.request import Request
+from snuba.request.request_settings import RequestSettings
```
Is sorting here due to circular imports?

fpacifici replied:
No. It is due to codemod. After codemodding everything to logical and physical I did not notice the order was not right anymore.

fpacifici replied:
fixed

tkaemming reviewed (commented):
This looks good to me so far aside from the updates to rename `StorageQueryPlanBuilder` and `StorageQueryPlan` (as long as I understood correctly that these are going to change to be refer to the physical query model.) The docstring for `Dataset` should probably be updated as well to reflect the logical/physical terminology. (And this PR title.)

fpacifici commented on snuba/query/processors/physical.py:
```diff
@@ -0,0 +1,18 @@
+
+class QueryProcessor(ABC):
+    """
+    This class represents the same concept as the Logical QueryProcessor (see
+    query.processors.logical).
+    The only difference is that it works on the Physical Query instead of the
+    Logical Query.
```
I was thinking about whether to do that, I did not because I could not find an actual use for the generic. These are two disjoint classes (logical and physical) that just happen to have the same interface conceptually. But they are never used interchangeably (that would be really wrong), and they are fully abstract so they do not have any parametric behavior.
I am open to adding the generic, what do you think ?

tkaemming replied:
If it is more coincidental than intentional, I agree that it probably doesn't make much sense to make it generic. At the same time, I'd probably avoid implying any intentional relationship to the logical query in the docstring as well (rather than explaining its behavior in relation to a different class, it should probably just describe it on its own terms.) It might also avoid any confusion down the line if the logical query processor changes for some reason.

fpacifici replied:
That makes sense, will rewrite the docstring

fpacifici replied:
Docstring is rewritten now.

fpacifici commented on snuba/web/db_query.py:
```diff
@@ -57,8 +57,9 @@ def update_query_metadata_and_stats(
 
 def raw_query(
-    request: Request,
+    physical_query: Query,
     query: ClickhouseQuery,
```
Yeah, there are a few things we discussed yesterday that I did not yet have time to retrofit in this change.

About specifically passing two query objects. That pissed me off quite a bit. I tried to remove that duplication but:
1) if I pass only Query here, it means I need to bring over the logic that formats it. That is (for tracing reasons) in a context. Thus it means this whole method which is already unreadable, gets indented more.
2) if I pass only ClickhosueQuery it just does not work because this method depends on the physical query object.

I gave up only because this is not different than what we are already doing. The physical query is just embedded in the Rrequest object so we are already passing both.

tkaemming replied:
That's fine, I just couldn't remember if we actually discussed that and what the conclusion was if we did. As long as `ClickhouseQuery` eventually becomes a subtype of `PhysicalQuery` (or interchangeable with it) and we fix it at that point, that's fine with me. I just wanted to be sure that was actually the plan.

fpacifici replied:
now we have Clickhouse Query (query.clickhouse.Query)

fpacifici committed:
Merge branch 'master' into feat/split_query_class

fpacifici committed:
Turn ClickhouseQuery into Formatter

fpacifici committed:
Rename physical query to clickhouse query

fpacifici committed:
Rename physical processor to clickhouse processor

fpacifici committed:
Fix import orders

fpacifici committed:
Rename Formatter into SqlQuery

fpacifici committed:
Rename storage query plan to clickhouse

fpacifici committed:
Update comments

fpacifici committed:
Fix translator abstraction

fpacifici committed:
Fix naming around

fpacifici committed:
Clarify some comments and names

fpacifici committed:
Rename SqlQUery

fpacifici committed:
Rename formatted query

fpacifici committed:
Remove irrelevant comment

fpacifici commented on snuba/clickhouse/native.py:
```diff
@@ -219,8 +219,12 @@ def __transform_result(self, result, with_totals: bool) -> Result:
         self,
-        query: ClickhouseQuery,
-        # TODO: move Clickhouse specific arguments into DictClickhouseQuery
+        # TODO: After we remove DictSqlQuery (that does all query processing for the
```
I tried a few alternatives here:
- just pass the Clickhosue Query. Does not work because other parts of the query processor before depend on the formatted query and we cannot format the old query representation twice (the AST solves this) since the formatting process changes the original query (not kidding).
- call the SqlQuery a "formatter". That would feel confusing since we would pass a "formatter" and not a query to the reader.
- frame the SqlQuery as a wrapper for the Clickhosue Query that provides formatting.  This is where I landed.

I think we can live with this till the AST is done (specifically now that we are actively working towards that) so that we can treat formatting as the stateless process it is supposed to be, let the Reader call the formatter and remove all the dependencies on the formatted query before the Reader.

tkaemming replied:
Thanks for the explanation, this is really helpful.

fpacifici commented on snuba/datasets/dataset.py:
```diff
@@ -45,7 +46,7 @@ class Dataset(object):
       by the dataset
     - the Storage to run the query onto is selected and the query is transformed
-      into a Storage Query. This is done by a StorageQueryPlanBuilder. This object
+      into a Storage Query. This is done by a ClickhouseQueryPlanBuilder. This object
```
I made the dataset depend directly on ClickhouseQueryPlanBuilder instead of having an intermediate generic QueryPlanBuilder. If I made the dataset depend on an abstract query plan builder instead, the query processing logic (which is clickhosue specific) would have had to ensure the instance returned by the dataset is the clickhouse one with various isinstance. 
Considering we are not likely to have any non clickhouse storage any time soon (for quite some time), it seems that this dependency is less leaky after we decided physical layer = clickhouse for the foreseeable future.

Adding the abstract QueryPlanBuilder is an easy change. Let me know what you think.

fpacifici commented on snuba/query/processors/__init__.py:
```diff
@@ -0,0 +1,56 @@
+from abc import ABC, abstractmethod
```
This entire file is just moved from query/query_processors.py. There should be no change.

fpacifici commented:
Adapted to the discussion we had last week. This is the summary of the changes
- physical layer is now clickhosue specific. Names (I hope all of them) are adapted
- some naming refactoring for the old ClickhouseQuery which was really a query formatter
- addressed comments of the previous review

remark:
for the naming strategy I tended not to add `Clickhouse` to the names of the classes  that deal with the physical layer and put them in the clickhouse package instead. There should not be confusion since the places where we deal with logical query and clickhouse query at the same time are very few).  Opinions on them? If you find ClickhouseQuery more clear than clickhouse.query.Query and ClickhouseQueryProcessor better than clickhouse.processors.QueryProcessors I am open to change them.

fpacifici commented on snuba/datasets/plans/query_plan.py:
```diff
@@ -50,24 +45,25 @@ class QueryPlanExecutionStrategy(ABC):
 
 
-class StorageQueryPlanBuilder(ABC):
+class ClickhouseQueryPlanBuilder(ABC):
```
I think there is still a way we should pursue to preserve Clickhouse independence in some of these interfaces, but it requires to explicitly call out dataset and storages as clickhouse datasets and clickhouse storages:
- clikchouse query and clickhouse query plan stay as they are in this PR
- clickhouse query processor will have a Generic interface on top with TQueryType as parameter
- query translator will have a Generic interface on top with TQueryType
- clickhosue query plan builder will have a Generic interface on top (same parameter)
- storage will have a generic interface on top with the same parameter and all our storages will be clickhouse specific and extend from this
- dataset will be have a generic interface on top and all our dataset will be clickhouse specific.

This way we are adding some clickhosue agnostic interfaces to several components but, since we do not have a mediator component between clickhouse agnostic storage level and clickhosue specific storage level, everything from dataset down to storages have to call out clickhosue explicitly.

fpacifici replied:
Doing this here: https://github.com/getsentry/snuba/pull/904
The commit touches a lot of files which would make this PR even larger.

fpacifici committed:
Merge branch 'master' into feat/split_query_class

tkaemming commented on snuba/query/processors/events_column_processor.py:
```diff
@@ -1,12 +1,6 @@
-    ConditionFunctions,
-)
-from snuba.query.dsl import literals_tuple
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Query
```
I don't think `snuba.query.processors` is the right place for these physical query processors at this point.

I think this should actually move into the file containing the events dataset or storage:

1. it's very coupled with that physical schema
2. that schema is also coupled with ClickHouse
3. this seems to have very little reuse value (so it doesn't belong in `snuba.clickhouse`)

It probably makes sense to do this as a follow up to minimize this specific change, though.

More broadly speaking, I think in general we should start more seriously splitting the code that is reusable (dataset classes, processor interfaces, etc) from the code that isn't (dataset instances, processors subclasses, etc) since those boundaries are becoming more clearly defined. (There is probably a better way to describe this logical separation but I can't find the right words this morning.) Personally, I find it a bit disorienting when reviewing that the dataset instances (which are almost practically configuration) are defined next to the dataset "framework" itself at a sibling level for example.

fpacifici replied:
Yeah I agree we need to take a stab to a consistent reorganization of the code base. 
I could start by moving this file to the datasets/storages directory. That's where the events storage is.

tkaemming commented on snuba/query/processors/join_optimizers.py:
```diff
@@ -1,7 +1,7 @@
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Query
```
(In contrast to the previous comment, this one actually probably _does_ belong in `snuba.clickhouse`.)

tkaemming commented on snuba/query/processors/prewhere.py:
```diff
@@ -1,8 +1,8 @@
 from snuba import settings, util
-from snuba.query.query import Query
-from snuba.query.query_processor import QueryProcessor
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Query
```
(Ok, this one also probably belongs in `snuba.clickhouse` too.)

fpacifici replied:
Does it? I would have put together with the storages (datasets/storages)

tkaemming replied:
I think that the physical storage implementations probably belong in `snuba.clickhouse` too, so I don't think these are incompatible views. This also reflects my understanding of  https://github.com/getsentry/snuba/pull/880#discussion_r415955384, specifically this part:

> storage will have a generic interface on top with the same parameter and all our storages will be clickhouse specific and extend from this

fpacifici replied:
yeah, moving all the clickhouse storages into snuba/clickhouse makes sense.
Will see what I can move in this PR without loosing git history.

tkaemming replied:
I think it would be fine (better, really) to do this as a follow up change (as noted with https://github.com/getsentry/snuba/pull/880#discussion_r416777412) since this is one is pretty large as-is.

tkaemming commented on snuba/clickhouse/sql.py:
```diff
@@ -0,0 +1,35 @@
+    This class embeds the formatting logic that transforms a Clickhouse Query into a SQL string.
+    The way this formatting is performed depends on the implementations.
+
+    This abstraction should not be needed. What we would need is a stateless formatter that
+    transforms the Clickhouse Query into a string and that can be used directly by the reader.
+    As long as we have DictSqlQuery around, this is not achievable because,
+    in the old representation, query processing and formatting are mixed with each other
+    and, worse, they are stateful (we cannot format the query twice).
+    As a consequence, we ned to instantiate this class before we get to the reader and pass it
+    around so it guarantees formatting happens only once. Once that class is gone we can turn
+    this into a stateless formatter.
```
👍 (on the comment, not the necessity of this class)

tkaemming commented on snuba/clickhouse/native.py:
```diff
@@ -219,8 +219,12 @@ def __transform_result(self, result, with_totals: bool) -> Result:
+        # TODO: After we remove DictSqlQuery (that does all query processing for the
+        # legacy query representation) we should pass the Clickhouse Query here. and the reader
+        # should rely on the formatter to get a SQL string and, thus, be able to provide parameters
+        # to the formatter (like the format clause itself).
```
Thinking about it — this probably means that the `FORMAT` clause is _not_ part of `ClickhouseQuery` (since its only pertinent to the `Reader`?)

fpacifici replied:
it seems to me that the reader implementation is the only one who cares about the output format. At least I thought this was the reason and that it was tied to the HTTP Reader.
Was there any other reason to make format part of the Clickhosue Query ?

tkaemming replied:
> Was there any other reason to make format part of the Clickhosue Query ?

Not that I can think of. (This was more of a clarifying question than one that was supposed to imply a change in direction.)

tkaemming commented on snuba/web/db_query.py:
```diff
@@ -189,8 +212,12 @@ def execute_query_with_deduplication(
-    query: ClickhouseQuery,
+    clickhouse_query: Query,
+    request_settings: RequestSettings,
+    # TODO: Passing the SqlQuery (which basically wraps the query and formats it)
+    # will be needed as long as the legacy query representation will be around.
+    # See DictSqlQuery.
+    formatted_query: SqlQuery,
```
This comment is helpful but it might be better for this to happen earlier in the file? It's easy to miss when starting with `execute_query`. Or maybe this function should come first, or the comment could be in both places? I'm not sure what makes sense here.

fpacifici replied:
ok

fpacifici replied:
it is in both places now.

tkaemming reviewed (commented):
I think you're going to need a (potentially pretty substantial) rebase but overall this looks good.

fpacifici commented on snuba/clickhouse/processors.py:
```diff
@@ -0,0 +1,25 @@
+
+    @abstractmethod
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        # TODO: Make the Query class immutable.
```
Contrary to the docstring: well yes, the docstring reflects the current state, this is a TODO. When/if we do this the docstring will change accordingly.

I think immutability would still be useful, specifically with the AST (expressions are immutable already, so replacing conditions will work differently than today).
There are still some cases where I think side effects are dangerous, they materialize in tracing now, (we store the query in a span, then we mutate it and when the span is recorded we record the final value instead of the initial one). This is a relatively small cases but it shows the danger. 

I think we should revisit this point when we get there. I will turn the TODO into "consider making the query immutable"

tkaemming replied:
Alright, thanks for clarifying. To be clear: I wasn't necessarily _advocating_ for making the query mutable, just noting it doesn't seem to be as urgent of an issue as I thought it was previously.

> expressions are immutable already, so replacing conditions will work differently than today

Ah, good point.

fpacifici commented on snuba/clickhouse/query.py:
```diff
@@ -1,25 +1,5 @@
-        return query
+# This defines the type for the Clickhouse query that operates on storages.
+# Obviously this is supposed to become more complex than just a type alias.
+Query = LogicalQuery
```
I thought this way would already prevent the type from leaking outwards. Will double check and change to NewType.

fpacifici replied:
using NewType

fpacifici commented on snuba/datasets/discover.py:
```diff
@@ -15,17 +15,18 @@
 from snuba.datasets.dataset import TimeSeriesDataset
 from snuba.datasets.factory import get_dataset
 from snuba.datasets.plans.single_storage import SelectedStorageQueryPlanBuilder
+from snuba.datasets.plans.translators import QueryTranslator
```
older revision. Forgot to remove it.

fpacifici commented on snuba/datasets/plans/query_plan.py:
```diff
@@ -4,36 +4,31 @@
+    It embeds the Clickhouse Query (the query to run on the storage after translation),
+    and the sequence of storage specific QueryProcessors to apply to the query after
+    the the storage has been selected.
+    It also provides a plan execution strategy, that takes care of coordinating the
+    execution of the query against the database and eventually to split it into chunks.
```
ok

tkaemming commented:
> for the naming strategy I tended not to add `Clickhouse` to the names of the classes that deal with the physical layer and put them in the clickhouse package instead. There should not be confusion since the places where we deal with logical query and clickhouse query at the same time are very few). Opinions on them?

Oops, missed this comment earlier but I think this is a reasonable choice, no issues from me.

fpacifici committed:
Merge branch 'master' into feat/split_query_class

fpacifici commented on snuba/request/__init__.py:
```diff
@@ -5,7 +5,7 @@
 from typing import Any, Mapping
 
-from snuba.query.query import Query
+from snuba.query.logical import Query
 from snuba.request.request_settings import RequestSettings
```
these seem correct to me.

tkaemming replied:
Uhhhh, yeah, you're right. Sorry, don't know what I'm on about today.

fpacifici committed:
Addressing comments

fpacifici commented on snuba/datasets/plans/query_plan.py:
```diff
@@ -50,9 +46,14 @@ class QueryPlanExecutionStrategy(ABC):
         """
-        Executes the query plan. The request parameter provides query and query settings.
+        Executes the query plan.
         The runner parameters is a function to actually run one individual query on the
```
fixed

fpacifici commented on snuba/datasets/discover.py:
```diff
@@ -1,5 +1,5 @@
 from datetime import timedelta
-from typing import Mapping, Sequence
+from typing import Mapping, Sequence, Tuple
```
fixed

fpacifici committed:
Move storage coupled processors

fpacifici committed:
Various nits

tkaemming reviewed (approved):
This looks good, between this and and some of the other recent changes this is all starting to feel like it's getting close.
